shapely/coordinates.py

```diff
@@ -129,6 +142,127 @@ def transform(
+        A function that transforms a (N, 2) or (N, 3) ndarray of float64 to
+        another (N, 2) or (N, 3) ndarray of float64.
+        The function may change the value of N.
+    include_z : bool, optional, default False
+        If False, always return 2D geometries.
+        If True, the data being passed to the
+        transformation function will include the third dimension
+        (if a geometry has no third dimension, the z-coordinates
+        will be NaN). If None, will infer the dimensionality using
+        ``has_z``. Note that this inference
+        can be unreliable with empty geometries or NaN coordinates: for a
+        guaranteed result, it is recommended to specify ``include_z`` explicitly.
+    interleaved : bool, default True
+        If set to False, the transformation function should accept 2 or 3 separate
+        arguments (x, y and optional z) instead of a single one. The return value
+        must be a tuple of (x, y and optional z). The arguments and return values
+        are one-dimensional arrays, however, if a TypeError is raised by the
+        supplied transformation function, it will be called with scalar floats.
+
+    See Also
+    --------
+    has_z : Returns a copy of a geometry array with a function applied to its
+        coordinates.
+    transform : Transform arrays of Geometry objects.
+
+    Examples
+    --------
+    >>> from shapely import LineString, Point
+
+    Reduce a linestring to only its first 2 points:
+
+    >>> transform_resize(LineString([(2, 2), (4, 4), (6, 6)]), lambda coords: coords[:2])
+    <LINESTRING (2 2, 4 4)>
+
+    The same with a function that accepts separate x, y arrays:
+
+    >>> transform_resize(LineString([(2, 2), (4, 4), (6, 6)]), lambda x, y: (x[:2], y[:2]), interleaved=False)
+    <LINESTRING (2 2, 4 4)>
+
+    Transform a point using a lambda function that accepts only scalars:
+
+    >>> transform_resize(Point(0, 0), lambda x, y: (x + 1, y + 2), interleaved=False)
+    <POINT (1 2)>
+    """
+    if geom is None:
+        return geom
+    if geom.is_empty:
+        if include_z is True:
+            return shapely.force_3d(geom)
+        elif include_z is False:
+            return shapely.force_2d(geom)
+        else:
+            return geom
+    if include_z is None:
+        include_z = geom.has_z
+
+    def _transform_internal(simple_geom):
+        coords = get_coordinates(simple_geom, include_z=include_z)
+        if interleaved:
+            return transformation(coords)
+        else:
+            # A transformation that assumes x, y, z are single values will likely raise a
+            # TypeError, in which case we'll try again.
+            try:
+                return zip(*transformation(*coords.T))
+            except TypeError:
+                # A func that assumes x, y, z are single values will likely raise a
+                # TypeError, in which case we'll try again.
+                return [transformation(*c) for c in coords]
```

Comment:
Personally I don't really like this try/except attempt to "infer whether the function is array-wise or element-wise". 

We _could_ also consider providing that as two separate transform functions, if we think this is worth providing to the user, to make it a clear choice for the user (I think typically the user will always know / have control over what kind of function they have). 

Another reason making this mixture messy is that AFAIK the second one can actually never resize your coordinates? So the use case of "resize coordseq" and "scalar function" will also never overlap?

Reply:
I also dislike this way inferring - I kept it only to make the migration of existing users of `ops.transform` easy.

> Another reason making this mixture messy is that AFAIK the second one can actually never resize your coordinates? So the use case of "resize coordseq" and "scalar function" will also never overlap?

That is well observed :) so in that case the "scalar function" version would better fit the existing `shapely.transform` function? I could think of an `call_with_scalars` kwarg for that function.

Reply:
> so in that case the "scalar function" version would better fit the existing `shapely.transform` function? I could think of an `call_with_scalars` kwarg for that function.

I don't really like adding such a keyword, but at the same providing a third transform method is also not great. 

But a counterargument for a keyword to `transform` is that this might conflict/ be incompatible with other keywords (like interleaved)

Now, I assume we can leave that for another issue, and you are focusing this PR just on the new resize version?